src/main/proto/netflix/titus/titus_cluster_membership_api.proto

```diff
@@ -0,0 +1,205 @@
+message ServiceInstance {
+
+    enum ServiceState {
+
+        // Service instance is starting, and its operational state is not known yet.
+        Starting = 0;
+
+        // Service is inactive, and does not participate in the leader election process. Predefined
+        // reason codes in this state:
+        // * 'outOfService' - the instance is healthy, but taken of out of service by an administrator
+        // * 'unhealthy' - the instance is not healthy
+        Inactive = 1;
+
+        // Service is fully activated, and if does not participate in the leader election process, is ready to
+        // take traffic.
+        Active = 2;
+    }
+
+    enum LeadershipState {
+        // Service is healthy, and participates in the leader election process, but is not the leader yet.
+        NonLeader = 0;
+
+        // Service became a leader, and is activating all components. It does not accept client requests yet.
+        LeaderActivating = 1;
+
+        // Service is a fully activated leader.
+        LeaderActivated = 2;
+    }
+
+    string instanceId = 1;
+
+    string instanceGroupId = 2;
+
+    // (Required) Service instance state
+    ServiceState state = 3;
+
+    // (Required) For services with a leader, indicates the leadership status of this instance. Otherwise, always
+    // set to `NonLeader`.
+    LeadershipState leadershipState = 4;
+
+    string ipAddress = 5;
+
+    repeated ServicePort serverPorts = 6;
+
+    // Collection of key/value pairs. Read-only special name spaces:
+    // * 'aws.deployment.XXX' - deployment data (ASG creation time, EC2 instance launch time, etc)
+    // * 'aws.asgTags.XXX' - service's ASG tags
+    // * 'aws.ec2Tags.XXX' - service's EC2 tags
+    // * 'release.XXX' - release data (number, build time, bake time)
+    map<string, string> labels = 7;
+}
+
+message ClusterMembershipEvent {
+    message ServiceInstanceUpdate {
+        ChangeRecord change = 1;
+        ServiceInstance previous = 2;
+        ServiceInstance after = 3;
+    }
+
+    message ServiceInstanceRemoved {
+        ChangeRecord change = 1;
+        ServiceInstance last = 2;
+    }
+
+    oneof Event {
+        ServiceInstanceUpdate serviceUpdated = 1;
+        ServiceInstanceRemoved serviceRemoved = 2;
+    }
+}
+
+// ----------------------------------------------------------------------------
+// GRPC services
+
+message Id {
+    string id = 1;
+}
+
+message ServiceInstances {
+    repeated ServiceInstance instances = 1;
+}
+
+message ChangeLogRequest {
+    // Filtering criteria:
+    // * 'ids' - comma separated list of instance ids
+    // * 'instanceLimit' - maximum number of records to return per instance
+    // * 'limit' - maximum number of records to return
+    // * 'before' - events before the given timestamp (epoch)
+    // * 'since' - events at or after the given timestamp (epoch)
+    map<string, string> filters = 1;
+
+    string fields = 2;
+}
+
+message ChangeLogResponse {
+    message ChangeRecord {
```

Comment:
Should this be renamed to not conflict with the above ChangeRecord message? Maybe ChangeLogRecord?

Reply:
Good catch. I have renamed this to `Entry`.